packages/cli/src/utils/scaffold.js

```diff
@@ -0,0 +1,109 @@
+const path = require('path');
+const colors = require('colors/safe');
+const _ = require('lodash');
+
+const { ensureDir, fileExistsSync, readFile, writeFile } = require('./files');
+const { splitFileFromPath } = require('../utils/string');
+const { createRootRequire, addKeyToPropertyOnApp } = require('./ast');
+const { snakeCase } = require('./misc');
+
+const plural = type => (type === 'search' ? `${type}es` : `${type}s`);
+
+const getTemplatePath = actionType =>
+  path.join(__dirname, '../..', `scaffold/${actionType}.template.js`);
+
+const createTemplateContext = (action, noun, includeComments) => {
+  // if noun is "Cool Contact"
+  return {
+    ACTION: action, // trigger
+    ACTION_PLURAL: plural(action), // triggers
+
+    KEY: snakeCase(noun), // "cool_contact", the action key
+    NOUN: _.capitalize(noun), // "Cool contact", the noun
+    LOWER_NOUN: noun.toLowerCase(), // "cool contact", for use in comments
+    // resources need an extra line for tests to "just run"
+    MAYBE_RESOURCE: action === 'resource' ? 'list.' : '',
+    INCLUDE_INTRO_COMMENTS: includeComments
+  };
+};
+
+const writeTemplateFile = async (
+  actionType,
+  templateContext,
+  destinationPath,
+  preventOverwrite
+) => {
+  const templatePath = getTemplatePath(actionType);
+
+  if (preventOverwrite && fileExistsSync(destinationPath)) {
+    const [location, filename] = splitFileFromPath(destinationPath);
+
+    throw new Error(
+      [
+        `File ${colors.bold(filename)} already exists within ${colors.bold(
+          location
+        )}.`,
+        'You can either:',
+        '  1. Choose a different filename',
+        `  2. Delete ${filename} from ${location}`,
+        `  3. Run ${colors.italic('scaffold')} with ${colors.bold(
+          '--force'
+        )} to overwrite the current ${filename}`
+      ].join('\n')
+    );
+  }
+
+  const template = (await readFile(templatePath)).toString();
+  const renderTemplate = _.template(template);
+
+  await ensureDir(path.dirname(destinationPath));
+  await writeFile(destinationPath, renderTemplate(templateContext));
+};
+
+const updateEntryFile = async (
+  entryFilePath,
+  varName,
+  newFilePath,
+  actionType,
+  newActionKey
+) => {
+  let codeStr = (await readFile(entryFilePath)).toString();
+  const entryName = splitFileFromPath(entryFilePath)[1];
+  const relativePath = path.relative(path.dirname(entryFilePath), newFilePath);
+
+  try {
+    codeStr = createRootRequire(codeStr, varName, `./${relativePath}`);
+    codeStr = addKeyToPropertyOnApp(codeStr, plural(actionType), varName);
+    await writeFile(entryFilePath, codeStr);
+
+    // validate the edit happened correctly
+    // can't think of why it wouldn't, but it doesn't hurt to double check
+    const rewrittenIndex = require(entryFilePath);
+    if (!_.get(rewrittenIndex, [plural(actionType), newActionKey])) {
```

Comment:
When I test, I always see the "Oops, we could not reliably rewrite" message even if my `index.js` seems to be correctly edited. I think it's because `require(entryFilePath)` is cached, which would always give us the unmodified file content. To get around the cache, we should probably use `readFile` or find a way to force `require` to reload.

Reply:
interesting, outside of bugs mine has worked every time :thinking:

![](https://zappy.zapier.com/45691eaf8d18dd8da9c1cf1d1102e993.png)

I can clear the require cache though, to make sure we're getting a fresh copy.

Reply:
I re-tested and it wasn't the cache. It's because I haven't run `npm|yarn install` within the app directory, so when you `require(entryFilePath)`, it raises an error. But the entry was correctly inserted to `index.js`. I think it's because before the `require`, you only parse `index.js` syntactically, which doesn't require dependencies.

I wonder is there a way to double check the edit without requiring a developer run `npm/yarn install` beforehand? After all, when scaffolding, there are high chances the dependencies haven't been installed yet.

Reply:
I've been thinking about this. I think we can safely assume that when this is run via `init`, it will always succeed (no need to re-check). So we only need to worry about the case where a dev is adding it to an existing app. We could probably do logic that this depends on a valid app install (aka they've installed `node_modules`) if directly as a command. Either bail early or do the work and not check it. I sort of like the former, since that's how we can ensure it worked (and not leave code in a bad state).

Reply:
I like the former too: check if they've installed `node_modules` before we re-check.